src/nbl/video/COpenGLCommandBuffer.h

```diff
@@ -855,27 +854,21 @@ class COpenGLCommandBuffer final : public IGPUCommandBuffer
         return true;
     }
 
-    bool waitEvents(uint32_t eventCount, event_t** pEvents, std::underlying_type_t<asset::E_PIPELINE_STAGE_FLAGS> srcStageMask, std::underlying_type_t<asset::E_PIPELINE_STAGE_FLAGS> dstStageMask,
-        uint32_t memoryBarrierCount, const asset::SMemoryBarrier* pMemoryBarriers,
-        uint32_t bufferMemoryBarrierCount, const SBufferMemoryBarrier* pBufferMemoryBarriers,
-        uint32_t imageMemoryBarrierCount, const SImageMemoryBarrier* pImageMemoryBarriers
-    ) override
+    bool waitEvents(uint32_t eventCount, event_t** pEvents, const SDependencyInfo* depInfos) override
     {
+        if (eventCount == 0u)
+            return false;
         for (uint32_t i = 0u; i < eventCount; ++i)
             if (!this->isCompatibleDevicewise(pEvents[i]))
                 return false;
-        if (eventCount == 0u)
-            return false;
-        SCmd<impl::ECT_WAIT_EVENTS> cmd;
-        cmd.eventCount = eventCount;
-        auto* events = getGLCommandPool()->emplace_n<core::smart_refctd_ptr<event_t>>(eventCount);
-        if (!events)
-            return false;
         for (uint32_t i = 0u; i < eventCount; ++i)
-            events[i] = core::smart_refctd_ptr<event_t>(pEvents[i]);
-        cmd.events = events;
-        cmd.barrier = barriersToMemBarrierBits(memoryBarrierCount, pMemoryBarriers, bufferMemoryBarrierCount, pBufferMemoryBarriers, imageMemoryBarrierCount, pImageMemoryBarriers);
-        pushCommand(std::move(cmd));
+        {
+            SCmd<impl::ECT_WAIT_EVENTS> cmd;
+            cmd.event = core::smart_refctd_ptr<event_t>(pEvents[i]);
+            auto& dep = depInfos[i];
+            cmd.barrier = barriersToMemBarrierBits(dep.memBarrierCount, dep.memBarriers, dep.bufBarrierCount, dep.bufBarriers, dep.imgBarrierCount, dep.imgBarriers);
+            pushCommand(std::move(cmd));
```

Comment:
multiple `glMemoryBarrier` calls will be VERY hurtful (can't you just OR the barrier bits and join into a single cmd?)

Reply:
hm, right, haven't thought about this